velox/substrait/VeloxToSubstraitPlan.cpp

```diff
@@ -170,13 +176,144 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
+    google::protobuf::Arena& arena,
+    const std::shared_ptr<const core::AggregationNode>& aggregateNode,
+    ::substrait::AggregateRel* aggregateRel) {
+  // Process the source Node.
+  std::vector<core::PlanNodePtr> sources = aggregateNode->sources();
+  // Check there only have one input.
+  VELOX_USER_CHECK_EQ(
+      1, sources.size(), "Aggregation plan node must have exactly one source.");
+  const auto& source = sources[0];
+
+  // Build source.
+  toSubstrait(arena, source, aggregateRel->mutable_input());
+
+  // Convert aggregate grouping keys, such as: group by key1, key2.
+  auto inputType = source->outputType();
+  auto groupingKeys = aggregateNode->groupingKeys();
+  int64_t groupingKeySize = groupingKeys.size();
+  ::substrait::AggregateRel_Grouping* aggGroupings =
+      aggregateRel->add_groupings();
+
+  for (int64_t i = 0; i < groupingKeySize; i++) {
+    aggGroupings->add_grouping_expressions()->MergeFrom(
+        exprConvertor_->toSubstraitExpr(
+            arena,
+            std::dynamic_pointer_cast<const ITypedExpr>(groupingKeys.at(i)),
+            inputType));
+  }
+
+  // AggregatesSize should be equal to or greater than the aggregateMasks Size.
+  // Two cases: 1. aggregateMasksSize = 0, aggregatesSize > aggregateMasksSize.
+  // 2. aggregateMasksSize != 0, aggregatesSize = aggregateMasksSize.
+  auto aggregates = aggregateNode->aggregates();
+  auto aggregateMasks = aggregateNode->aggregateMasks();
+  int64_t aggregatesSize = aggregates.size();
+  int64_t aggregateMasksSize = aggregateMasks.size();
+  VELOX_CHECK_GE(aggregatesSize, aggregateMasksSize);
+
+  for (int64_t i = 0; i < aggregatesSize; i++) {
+    ::substrait::AggregateRel_Measure* aggMeasures =
+        aggregateRel->add_measures();
+
+    std::shared_ptr<const core::FieldAccessTypedExpr> aggMaskExpr =
+        aggregateMasks.at(i);
+    // Set substrait filter.
+    ::substrait::Expression* aggFilter = aggMeasures->mutable_filter();
+    if (aggMaskExpr.get()) {
+      aggFilter->MergeFrom(exprConvertor_->toSubstraitExpr(
+          arena,
+          std::dynamic_pointer_cast<const ITypedExpr>(aggMaskExpr),
+          inputType));
+    } else {
+      // Set null.
+      aggFilter = NULL;
+    }
+
+    // Process measure, eg:sum(a).
+    std::shared_ptr<const core::CallTypedExpr> aggregatesExpr =
+        aggregates.at(i);
+    ::substrait::AggregateFunction* aggFunction =
+        aggMeasures->mutable_measure();
+
+    // Aggregation function name.
+    std::string funName = aggregatesExpr->name();
+    // set aggFunction args.
+    std::vector<core::TypedExprPtr> aggregatesExprInputs =
+        aggregatesExpr->inputs();
+
+    int64_t aggregatesExprInputsSize = aggregatesExprInputs.size();
+    for (int64_t j = 0; j < aggregatesExprInputsSize; j++) {
+      ::substrait::Expression* aggFunctionExpr = aggFunction->add_args();
+      core::TypedExprPtr expr = aggregatesExprInputs[j];
+      // In velox plan, expr can only be the FieldAccessTypedExpr.
```

Comment:
This comment doesn't match the code. Let't cast to Field and fail if result is null.

Reply:
what about count(1),  it should be constantTypedExpe, right?

Reply:
change the comment to " // If the expr is CallTypedExpr, people need do project firstly."  Is this ok?